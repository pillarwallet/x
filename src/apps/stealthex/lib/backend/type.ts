export type UnavailableResponse = {
    state: 'unknown';
  };
  
export type FailedResponse<T> = {
    state: 'failed';
    status: number;
    message?: T;
};

export type ThrowableError<T> = UnavailableResponse | FailedResponse<T>;

export type ErrorResponse<T> = ThrowableError<T> & { data: null };

export type ThrowableResponse<T> = {
    status: number;
    data: T;
    headers: Headers;
};

export type SuccessResponse<T> = ThrowableResponse<T> & {
state: 'success';
};
  
export type Response<D, E> = SuccessResponse<D> | ErrorResponse<E>;

export type RequestResponse<
  T extends boolean,
  D = unknown,
  E = string,
> = T extends true ? ThrowableResponse<D> : Response<D, E>;