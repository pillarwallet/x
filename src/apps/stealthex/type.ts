import { CurrencyData } from './lib/backend/api';

export type Config = {
    amount: string;
    fixed: boolean;
    fiat: boolean;
    sendCurrency: CurrencyData;
    receiveCurrency: CurrencyData;
    getPairs?: CurrencyData[];
    sendPairs?: CurrencyData[];
    recepientAddress?: string;
    lockReceiveCurrency?: boolean;
}

export type ExchangeInfoCurrencyPair = {
    [key: string]: CurrencyData | undefined;
  };

export type ExchangeInfo = {
    id: string;
    type: string;
    timestamp: string;
    updated_at: string;
    currency_from: string;
    currency_to: string;
    amount_from: string;
    expected_amount: string;
    amount_to: string;
    amount_to_btc: string | null;
    address_from: string;
    address_to: string;
    extra_id_from: string | null;
    extra_id_to: string | null;
    tx_from: string | null;
    tx_to: string | null;
    status:
      | 'waiting'
      | 'confirming'
      | 'exchanging'
      | 'sending'
      | 'finished'
      | 'failed'
      | 'refunded'
      | 'expired'
      | 'verifying';
    refund_address: string | null;
    refund_extra_id: string | null;
    currencies: ExchangeInfoCurrencyPair;
    review?: boolean;
    payment_id?: string;
    mercuryo_signature: string;
};

export type StructuredCurrency = {
  fiat: CurrencyData[];
  popular: CurrencyData[];
  other: CurrencyData[];
};