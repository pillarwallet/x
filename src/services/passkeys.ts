import {
  startRegistration,
  startAuthentication,
} from '@simplewebauthn/browser';

// Base URL for Firebase functions - adjust this based on your deployment
const FIREBASE_FUNCTIONS_BASE_URL = 'http://localhost:5000/pillarx-staging/us-central1';

// Helper function to make API calls to Firebase functions
async function callFirebaseFunction(endpoint: string, data: any) {
  const url = `${FIREBASE_FUNCTIONS_BASE_URL}/passkeys/${endpoint}`;
  console.log(`Calling Firebase function: ${url}`, data);
  
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  console.log(`Response status: ${response.status}`);
  
  if (!response.ok) {
    const errorData = await response.json();
    console.error('Firebase function error:', errorData);
    throw new Error(errorData.error || `HTTP error! status: ${response.status}`);
  }

  const result = await response.json();
  console.log(`Firebase function success:`, result);
  return result;
}

// Register a new passkey for a user
export async function registerPasskey(userId: string, username: string): Promise<boolean> {
  try {
    // Step 1: Get registration options from server
    const options = await callFirebaseFunction('register/options', {
      userId,
      username,
    });

    // Step 2: Start registration on the client
    const credential = await startRegistration(options);

    // Step 3: Verify registration with server
    const verification = await callFirebaseFunction('register/verify', {
      userId,
      credential,
    });

    return verification.verified === true;
  } catch (error) {
    console.error('Passkey registration failed:', error);
    throw error;
  }
}

// Authenticate with an existing passkey
export async function authenticateWithPasskey(userId: string): Promise<boolean> {
  try {
    // Step 1: Get authentication options from server
    const options = await callFirebaseFunction('authenticate/options', {
      userId,
    });

    // Step 2: Start authentication on the client
    const credential = await startAuthentication(options);

    // Step 3: Verify authentication with server
    const verification = await callFirebaseFunction('authenticate/verify', {
      userId,
      credential,
    });

    return verification.verified === true;
  } catch (error) {
    console.error('Passkey authentication failed:', error);
    throw error;
  }
}

// Sign a custom challenge with passkey
export async function signWithPasskey(userId: string, customChallenge: string): Promise<{ verified: boolean; signedChallenge?: string; signature?: string }> {
  try {
    // Step 1: Get signing options from server with custom challenge
    const options = await callFirebaseFunction('sign/options', {
      userId,
      customChallenge,
    });

    // Step 2: Start authentication on the client (same as signing)
    const credential = await startAuthentication(options);

    // Step 3: Verify signing with server
    const verification = await callFirebaseFunction('sign/verify', {
      userId,
      credential,
    });

    return {
      verified: verification.verified === true,
      signedChallenge: verification.signedChallenge,
      signature: verification.signature,
    };
  } catch (error) {
    console.error('Passkey signing failed:', error);
    throw error;
  }
}

// Check if passkeys are supported in the current browser
export function isPasskeySupported(): boolean {
  return window.PublicKeyCredential !== undefined;
}

// Check if the current platform supports passkeys
export async function isPasskeyAvailable(): Promise<boolean> {
  if (!isPasskeySupported()) {
    return false;
  }

  try {
    return await PublicKeyCredential.isUserVerifyingPlatformAuthenticatorAvailable();
  } catch (error) {
    console.error('Error checking passkey availability:', error);
    return false;
  }
} 