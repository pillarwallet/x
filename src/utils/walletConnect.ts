import { SessionTypes } from '@walletconnect/types';
import { isAddress, isAddressEqual } from 'viem';

export const PERSONAL_SIGN = 'personal_sign';
export const ETH_SIGN = 'eth_sign';
export const ETH_SEND_TX = 'eth_sendTransaction';
export const ETH_SIGN_TX = 'eth_signTransaction';
export const ETH_SIGN_TYPED_DATA = 'eth_signTypedData';
export const ETH_SIGN_TYPED_DATA_V4 = 'eth_signTypedData_v4';
export const ETH_SEND_RAW_TRANSACTION = 'eth_sendRawTransaction';
export const ETH_ACCOUNTS = 'eth_accounts';
export const WALLET_SWITCH_CHAIN = 'wallet_switchEthereumChain';

export const WALLETCONNECT_EVENT = {
  SESSION_DELETE: 'session_delete',
  SESSION_PROPOSAL: 'session_proposal',
  SESSION_REQUEST: 'session_request',
  PROPOSAL_EXPIRE: 'proposal_expire',
  SESSION_REQUEST_EXPIRE: 'session_request_expire',
};

export const getWalletAddressesFromSession = (
  session: SessionTypes.Struct
): string[] => {
  const accounts: string[] = session.namespaces?.eip155?.accounts || [];
  return accounts.map((acc) => acc.split(':')[2]);
};

export const isAddressInSessionViaPrivy = (
  session: SessionTypes.Struct,
  EOAAddress: string | null | undefined
): boolean => {
  if (!EOAAddress || !isAddress(EOAAddress, { strict: false })) {
    return false;
  }

  const addresses = getWalletAddressesFromSession(session);

  return addresses.some((addr) => {
    if (!addr || !isAddress(addr, { strict: false })) {
      return false; // skip if the address is invalid
    }
    return isAddressEqual(addr as `0x${string}`, EOAAddress as `0x${string}`);
  });
};
