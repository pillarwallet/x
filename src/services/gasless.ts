/* eslint-disable import/extensions */
import { constants } from 'ethers';

import { isPolygonAssetNative } from '../utils/blockchain';
import { Token } from './tokensData';

export type Paymasters = {
  gasToken: string;
  chainId: number;
  epVersion: string;
  paymasterAddress: string;
};

interface ChainBalance {
  chainId: string;
  balance: string;
}

export const GasConsumptions = {
  native: 510000,
  native_arb: 910000,
  token: 550000,
  token_arb: 960000,
  nft: 630000,
  nft_arb: 1050000,
};

export const getAllGaslessPaymasters = async (
  chainId: number,
  tokens_list: Token[]
): Promise<Paymasters[] | null> => {
  try {
    const res = await fetch(
      `${import.meta.env.VITE_PAYMASTER_URL}/getAllCommonERC20PaymasterAddress`,
      {
        method: 'POST',
        body: JSON.stringify({}),
      }
    );
    const data = await res.json();
    if (data.message) {
      let paymasterObject = JSON.parse(data.message);
      paymasterObject = paymasterObject.filter(
        (item: { epVersion: string; chainId: number; gasToken: string }) =>
          item.epVersion === 'EPV_07' &&
          item.chainId === chainId &&
          tokens_list.find(
            (token: Token) =>
              token.contract === item.gasToken.toLowerCase() ||
              (isPolygonAssetNative(item.gasToken, item.chainId) &&
                token.contract === constants.AddressZero)
          )
      );
      return paymasterObject;
    }
    return null;
  } catch (err) {
    console.error(err);
    return null;
  }
};

export const getGasTankBalance = async (
  walletAddress: string
): Promise<number | null> => {
  try {
    const response = await fetch(
      `${import.meta.env.VITE_PAYMASTER_URL}/getGasTankBalance?sender=${walletAddress}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch gas tank balance: ${response.status}`);
    }

    const data = await response.json();
    const balances: ChainBalance[] = Object.values(data.balance || {});
    // Get cumalative balance for all chains
    const total = balances.reduce((sum, chainBalance) => {
      const balance = parseFloat(chainBalance.balance) || 0;
      return sum + balance;
    }, 0);
    return total;
  } catch (error) {
    console.error('Error fetching gas tank balance:', error);
    return null;
  }
};
