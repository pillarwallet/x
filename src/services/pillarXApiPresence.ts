// Need to use the React-specific entry point to import createApi
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// utils
import { CompatibleChains } from '../utils/blockchain';

export const pillarXApiPresence = createApi({
  reducerPath: 'pillarXApiPresence',
  baseQuery: fetchBaseQuery({
    baseUrl:
      process.env.REACT_APP_USE_TESTNETS === 'true'
        ? 'https://activity-nubpgwxpiq-uc.a.run.app'
        : 'https://activity-7eu4izffpa-uc.a.run.app',
  }),
  endpoints: (builder) => ({
    recordPresence: builder.mutation({
      query: (payload = {}) => {
        const chainIds =
          process.env.REACT_APP_USE_TESTNETS === 'true'
            ? [11155111]
            : CompatibleChains.map((chain) => chain.chainId);
        const chainIdsQuery = chainIds.map((id) => `chainIds=${id}`).join('&');

        return {
          url: '/',
          method: 'POST',
          params: {
            testnets: process.env.REACT_APP_USE_TESTNETS || 'true',
            chainIds: chainIdsQuery,
          },
          body: payload,
        };
      },
    }),
  }),
});

export const { useRecordPresenceMutation } = pillarXApiPresence;
