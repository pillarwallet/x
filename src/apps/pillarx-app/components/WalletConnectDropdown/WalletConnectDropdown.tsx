/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable react/no-unstable-nested-components */
import { CircularProgress } from '@mui/material';
import { useEffect, useRef, useState } from 'react';

// services
import { useWalletConnect } from '../../../../services/walletConnect';

// images
import ArrowDown from '../../images/arrow-down.svg';
import SettingIcon from '../../images/setting-wheel.svg';
import Ticked from '../../images/tick-square-ticked.svg';
import Unticked from '../../images/tick-square-unticked.svg';
import WalletConnectLogo from '../../images/wallet-connect-logo.svg';

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import SwitchToggle from '../SwitchToggle/SwitchToggle';
import Body from '../Typography/Body';
import BodySmall from '../Typography/BodySmall';

const WalletConnectDropdown = () => {
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const [clipboardToggle, setClipboardToggle] = useState<boolean>(false);
  const [showPermission, setShowPermission] = useState<boolean>(true);
  const [isSettingsView, setIsSettingsView] = useState<boolean>(false);
  const [isBrowserPermission, setIsBrowserPermission] = useState<boolean>(true);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const {
    connect,
    disconnect,
    disconnectAllSessions,
    activeSessions,
    isLoadingDisconnect,
    isLoadingDisconnectAll,
  } = useWalletConnect();

  const getClipboardText = async (prompt: boolean) => {
    try {
      return await navigator.clipboard.readText();
    } catch (e) {
      if (prompt) {
        // eslint-disable-next-line no-alert
        const walletConnectPrompt = window.prompt(
          'Enter the WalletConnect link here, or change your clipboard permissions in your browser settings.'
        );
        return walletConnectPrompt;
      }
      return false;
    }
  };

  const handleClipboardMessageClick = () => {
    setShowPermission(!showPermission);
  };

  const handleConnect = async () => {
    const clipboardText = await getClipboardText(true);
    if (clipboardText) {
      await connect(clipboardText);
    }
  };

  const handleDropdownToggle = () => {
    setIsDropdownOpen((prev) => !prev);
    setIsSettingsView(false);
  };

  const handleClipboardToggle = (checked: boolean) => {
    setClipboardToggle(checked);
  };

  const handleClipboardPermission = async () => {
    try {
      const clipboardPermission = await navigator.permissions.query({
        name: 'clipboard-read' as PermissionName,
      });

      if (clipboardPermission.state === 'granted') {
        handleClipboardToggle(true);
      } else {
        handleClipboardToggle(false);
      }

      clipboardPermission.addEventListener('change', () => {
        if (clipboardPermission.state === 'granted') {
          handleClipboardToggle(true);
        } else {
          handleClipboardToggle(false);
        }
      });
    } catch (e) {
      localStorage.setItem('CLIPBOARD_MESSAGE', 'false');
      setIsBrowserPermission(false);
    }
  };

  useEffect(() => {
    handleClipboardPermission();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsDropdownOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [dropdownRef]);

  const handleCompleteSetup = () => {
    localStorage.setItem('CLIPBOARD_MESSAGE', showPermission.toString());
    setIsDropdownOpen(false);
  };

  const numberActiveSessions =
    !!activeSessions && Object.entries(activeSessions).length;

  const DisplayContentWalletConnect = () => {
    if (
      localStorage.getItem('CLIPBOARD_MESSAGE') === 'true' ||
      isSettingsView
    ) {
      return (
        <>
          <div className="flex flex-col">
            {!isSettingsView && (
              <BodySmall className="font-semibold mb-3">Setup</BodySmall>
            )}
            <div className="flex justify-between items-center mb-1">
              <BodySmall className="text-xs font-semibold">
                Clipboard permission
              </BodySmall>
              <SwitchToggle
                isChecked={clipboardToggle}
                isClipboardSwitch
                onClick={() => getClipboardText(false)}
              />
            </div>
            <BodySmall className="text-xs text-purple_light">
              dApps can be automatically connected by copying the WalletConnect
              URL to clipboard.
            </BodySmall>
          </div>
          {!isSettingsView && (
            <div className="flex">
              <img
                src={showPermission ? Unticked : Ticked}
                alt="ticked-clipboard-permission"
                className="h-3.5 w-3.5 cursor-pointer mr-2"
                onClick={handleClipboardMessageClick}
              />
              <BodySmall className="text-xs text-purple_light">
                I understand, don&apos;t show this message again.
              </BodySmall>
            </div>
          )}
          <button
            type="button"
            onClick={() =>
              isSettingsView ? setIsSettingsView(false) : handleCompleteSetup()
            }
            className="flex p-1.5 items-center justify-center bg-purple_medium rounded-md w-full"
          >
            <BodySmall>
              {isSettingsView ? 'Close' : 'Completed set-up'}
            </BodySmall>
          </button>
        </>
      );
    }

    if (
      localStorage.getItem('CLIPBOARD_MESSAGE') === 'false' &&
      numberActiveSessions &&
      numberActiveSessions > 0
    ) {
      return (
        <div className="flex flex-col gap-4">
          <BodySmall className="text-xs text-purple_light text-center">
            Copy WalletConnect URL & click below to automatically connect
          </BodySmall>
          <div className="flex justify-between items-center gap-2">
            <button
              type="button"
              className="bg-purple_medium text-sm font-medium rounded-md w-full p-1.5"
              onClick={handleConnect}
            >
              Connect dApp
            </button>
            {isBrowserPermission && (
              <button
                type="button"
                className="flex items-center justify-center h-8 w-10 rounded-md border border-purple_light/[.1]"
                onClick={() => setIsSettingsView(true)}
              >
                <img
                  src={SettingIcon}
                  alt="wallet-connect-setting"
                  className="w-4 h-4"
                />
              </button>
            )}
          </div>
          {numberActiveSessions > 1 && (
            <div className="flex justify-between items-center mb-4">
              <BodySmall className="text-xs text-purple_light font-semibold">
                {numberActiveSessions} Connected
              </BodySmall>
              <button
                type="button"
                className="text-[11px] px-2 py-1.5 bg-container_grey border border-container_grey rounded-md"
                onClick={() => disconnectAllSessions}
              >
                {isLoadingDisconnectAll ? (
                  <CircularProgress
                    size={20}
                    sx={{ color: 'white' }}
                    data-testid="circular-loading"
                  />
                ) : (
                  'Disconnect all'
                )}
              </button>
            </div>
          )}

          <ul className="flex flex-col gap-4">
            {Object.values(activeSessions ?? {}).map((session: any) => (
              <li key={session.topic}>
                <div className="flex items-center justify-between">
                  <div className="flex items-center">
                    {session.peer.metadata.icons[0] ? (
                      <img
                        src={session.peer.metadata.icons[0]}
                        alt="dApp-logo"
                        className="w-[30px] h-[30px] object-fill rounded-full mr-2"
                      />
                    ) : (
                      <div className="w-[50px] h-[50px] object-fill rounded mr-3.5 overflow-hidden">
                        <RandomAvatar name={session.peer.data.name || ''} />
                      </div>
                    )}
                    <BodySmall className="line-clamp-1">
                      {session.peer.metadata.name}
                    </BodySmall>
                  </div>
                  <button
                    type="button"
                    className="text-[11px] px-2 py-1.5 border border-purple_light/[.1] rounded-md"
                    onClick={() => disconnect(session.topic)}
                  >
                    {isLoadingDisconnect ? (
                      <CircularProgress
                        size={20}
                        sx={{ color: 'white' }}
                        data-testid="circular-loading"
                      />
                    ) : (
                      'Disconnect'
                    )}
                  </button>
                </div>
              </li>
            ))}
          </ul>
        </div>
      );
    }

    if (
      localStorage.getItem('CLIPBOARD_MESSAGE') === 'false' &&
      (!numberActiveSessions || numberActiveSessions === 0)
    ) {
      return (
        <div className="flex flex-col gap-4">
          <BodySmall className="text-xs text-purple_light text-center">
            Copy WalletConnect URL & click below to automatically connect
          </BodySmall>
          <div className="flex justify-between items-center gap-2">
            <button
              type="button"
              className="bg-purple_medium text-sm font-medium rounded-md w-full p-1.5"
              onClick={handleConnect}
            >
              Connect dApp
            </button>
            {isBrowserPermission && (
              <button
                type="button"
                className="flex items-center justify-center h-8 w-10 rounded-md border border-purple_light/[.1]"
                onClick={() => setIsSettingsView(true)}
              >
                <img
                  src={SettingIcon}
                  alt="wallet-connect-setting"
                  className="w-4 h-4"
                />
              </button>
            )}
          </div>
          <BodySmall className="text-xs text-purple_light">
            No dApp connected yet.
          </BodySmall>
        </div>
      );
    }

    return null;
  };

  return (
    <div
      className="relative inline-block w-[254px] h-12 mb-4"
      ref={dropdownRef}
    >
      <div
        className={`flex w-full h-full items-center bg-medium_grey cursor-pointer ${isDropdownOpen ? 'rounded-t-md' : 'rounded-md'}`}
        onClick={handleDropdownToggle}
      >
        <div className="flex w-full gap-2 h-full items-center px-3 border-r-2 border-purple_light/[.1]">
          <img
            src={WalletConnectLogo}
            alt="wallet-connect-logo"
            className="w-[24px]"
          />
          <Body className="text-white">WalletConnect</Body>
          {numberActiveSessions ? (
            <div className="flex w-[19px] h-[19px] bg-purple_medium rounded justify-center items-center">
              <p className="font-semibold text-white text-[10px]">
                {numberActiveSessions}
              </p>
            </div>
          ) : null}
        </div>
        <div className="flex min-w-[46px] h-full items-center justify-center">
          <img
            src={ArrowDown}
            alt="arrow-down"
            className={`w-3.5 transform ${isDropdownOpen ? 'rotate-180' : ''}`}
          />
        </div>
      </div>
      {isDropdownOpen && (
        <div className="flex flex-col gap-4 absolute left-0 w-full p-4 bg-medium_grey rounded-b-md border-t-2 border-purple_light/[.1] z-10">
          <DisplayContentWalletConnect />
        </div>
      )}
    </div>
  );
};

export default WalletConnectDropdown;
