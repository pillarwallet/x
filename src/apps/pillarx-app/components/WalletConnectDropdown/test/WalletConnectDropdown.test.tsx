import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import type { Mock } from 'vitest';

// hooks
import { useWalletConnect } from '../../../../../services/walletConnect';

// components
import WalletConnectDropdown from '../WalletConnectDropdown';

vi.mock('../../../../../services/walletConnect', () => ({
  useWalletConnect: vi.fn(),
}));

Object.defineProperty(navigator, 'clipboard', {
  value: {
    readText: vi.fn(),
  },
  writable: true,
});

// 'Type' for the return value of `useWalletConnect` hook
interface UseWalletConnectProps {
  connect: vi.mock;
  disconnect: vi.mock;
  disconnectAllSessions: vi.mock;
  activeSessions:
    | Record<
        string,
        { topic: string; peer: { metadata: { name: string; icons: string[] } } }
      >
    | undefined;
  isLoadingConnect: boolean;
  isLoadingDisconnect: boolean;
  isLoadingDisconnectAll: boolean;
}

describe('<WalletConnectDropdown />', () => {
  const mockConnect = vi.fn();
  const mockDisconnect = vi.fn();
  const mockDisconnectAllSessions = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    (useWalletConnect as Mock).mockReturnValue({
      connect: mockConnect,
      disconnect: mockDisconnect,
      disconnectAllSessions: mockDisconnectAllSessions,
      activeSessions: {},
      isLoadingConnect: false,
      isLoadingDisconnect: false,
      isLoadingDisconnectAll: false,
    });
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer.create(<WalletConnectDropdown />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggles the dropdown on click', () => {
    render(<WalletConnectDropdown />);
    const dropdown = screen.getByText('WalletConnect');
    fireEvent.click(dropdown);
    expect(
      screen.getByText(
        'Copy WalletConnect URL & click below to automatically connect'
      )
    ).toBeInTheDocument();
    fireEvent.click(dropdown);
    expect(
      screen.queryByText(
        'Copy WalletConnect URL & click below to automatically connect'
      )
    ).not.toBeInTheDocument();
  });

  it('calls connect function with a URI when connect button is clicked', async () => {
    const copiedUri = 'some-mocked-uri';

    (navigator.clipboard.readText as Mock).mockImplementation(() =>
      Promise.resolve(copiedUri)
    );

    render(<WalletConnectDropdown />);
    fireEvent.click(screen.getByText('WalletConnect'));
    const connectButton = screen.getByText('Connect dApp');

    fireEvent.click(connectButton);

    await waitFor(() => {
      expect(mockConnect).toHaveBeenCalled();
    });
  });

  it('displays active sessions when available', () => {
    (useWalletConnect as Mock).mockReturnValue({
      connect: mockConnect,
      disconnect: mockDisconnect,
      disconnectAllSessions: mockDisconnectAllSessions,
      activeSessions: {
        session1: {
          topic: '123',
          peer: { metadata: { name: 'Test dApp', icons: [''] } },
        },
      },
      isLoadingConnect: false,
      isLoadingDisconnect: false,
      isLoadingDisconnectAll: false,
    });

    render(<WalletConnectDropdown />);
    fireEvent.click(screen.getByText('WalletConnect'));
    expect(screen.getByText('Test dApp')).toBeInTheDocument();
  });

  it('calls disconnect when disconnect button is clicked', async () => {
    (useWalletConnect as Mock<UseWalletConnectProps>).mockReturnValue({
      connect: mockConnect,
      disconnect: mockDisconnect,
      disconnectAllSessions: mockDisconnectAllSessions,
      activeSessions: {
        session1: {
          topic: '123',
          peer: { metadata: { name: 'Test dApp', icons: [''] } },
        },
      },
      isLoadingConnect: false,
      isLoadingDisconnect: false,
      isLoadingDisconnectAll: false,
    });
    render(<WalletConnectDropdown />);
    fireEvent.click(screen.getByText('WalletConnect'));
    const disconnectButton = screen.getByText('Disconnect');
    fireEvent.click(disconnectButton);
    await waitFor(() => {
      expect(mockDisconnect).toHaveBeenCalledWith('123');
    });
  });

  it('disconnectAll button appears when several sessions are connected', async () => {
    (useWalletConnect as Mock).mockReturnValue({
      connect: mockConnect,
      disconnect: mockDisconnect,
      disconnectAllSessions: mockDisconnectAllSessions,
      activeSessions: {
        session1: {
          topic: '123',
          peer: { metadata: { name: 'Test dApp', icons: [''] } },
        },
        session2: {
          topic: '456',
          peer: { metadata: { name: 'Another dApp', icons: [''] } },
        },
      },
      isLoadingConnect: false,
      isLoadingDisconnect: false,
      isLoadingDisconnectAll: false,
    });

    render(<WalletConnectDropdown />);
    fireEvent.click(screen.getByText('WalletConnect'));
    const disconnectAllButton = screen.getByText('Disconnect all');

    expect(disconnectAllButton).toBeInTheDocument();
  });
});
