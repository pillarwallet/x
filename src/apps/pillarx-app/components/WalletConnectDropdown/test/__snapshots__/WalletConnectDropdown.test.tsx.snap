// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WalletConnectDropdown /> renders correctly and matches snapshot 1`] = `
<div
  className="relative inline-block w-[254px] h-12 mb-4"
  id="walletConnect-dropdown"
>
  <div
    className="flex w-full h-full items-center bg-medium_grey cursor-pointer rounded-md"
    onClick={[Function]}
  >
    <div
      className="flex w-full gap-2 h-full items-center px-3 border-r-2 border-purple_light/[.1]"
    >
      <img
        alt="wallet-connect-logo"
        className="w-[24px]"
        src={
          {
            "ReactComponent": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "default": "wallet-connect-logo.svg",
          }
        }
      />
      <p
        className="text-base font-medium text-white"
      >
        WalletConnect
      </p>
    </div>
    <div
      className="flex min-w-[46px] h-full items-center justify-center"
    >
      <img
        alt="arrow-down"
        className="w-3.5 transform "
        src={
          {
            "ReactComponent": {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "default": "arrow-down.svg",
          }
        }
      />
    </div>
  </div>
</div>
`;
